File: Tree/binary_tree_to_dll.rb
Node = Struct.new(:key, :left, :right)
module BinaryTree
  module_function

  def inorder(node, prev)
    return if node == nil

    inorder(node.left, prev)
    # print "#{node.key}.prev(l) = #{store.curr.key} .. "
    node.left = prev[0]
    # print "#{store.curr.key}.next(r) = #{node.key} .. "
    prev[0].right = node
    # print "curr is #{store.curr.key}\n"
    prev[0] = node
    inorder(node.right, prev)

  end

  def to_dll(root)
    dummy = Node.new('dummy')
    store = [dummy]
    inorder(root, store)
    dummy.right.left = nil
    return dummy.right
  end

end

def print_list(root)
  while root != nil
    # puts "left: #{curr.left.key if curr.left} .. key: #{curr.key} .. right #{curr.right.key if curr.right}"
    print "#{root.key} "
    root = root.right
  end
  puts
end

root = Node.new(10)
root.left = Node.new(12)
root.right = Node.new(15)
root.left.left = Node.new(25)
root.left.right = Node.new(30)
root.right.left = Node.new(36)

head = BinaryTree.to_dll(root)

print_list(head)

File: Tree/binary_tree_to_dll.py
class Node:
    """ Class Node """
    def __init__(self, value):
        self.left = None
        self.data = value
        self.right = None

def bToDLL(root):
    dummy = Node("dummy")
    prev = [dummy]
    inorder(root, prev)
    head = dummy.right
    head.left = None
    return head

def inorder(node, prev):
    if node is None:
        return
    inorder(node.left, prev)

    node.left = prev[0]
    prev[0].right = node
    prev[0] = node

    inorder(node.right, prev)


def printInOrder(root):
    if root is None:
        return
    printInOrder(root.left)
    print(root.data, end = ' ')
    printInOrder(root.right)

def printLinear(head):
    while head != None:
        print(head.data, end = ' ')
        head = head.right
    print()


root = Node(10)
root.left = Node(12)
root.right = Node(15)
root.right.left = Node(36)
root.left.left = Node(25)
root.left.right = Node(30)

print("Tree inorder: ")
printInOrder(root)
print()

head = bToDLL(root)

print("DLL: ")
printLinear(head)
